Tighten up message size sanity checking

gavinandresen opened the pull request:
These commits tighten up message size sanity checking by:
- Adding a new signal (SanityCheckMessages) to the networking code
- Refactoring MAX_PROTOCOL_MESSAGE_SIZE from the networking code to main
- Adding additional checks for several messages (e.g. ping messages are at most 8 bytes) to the main signal handler (next to the code that handles each message type)

Extending the SanityCheckMessages signal handler (or adding an additional handler) to do more sophisticated DoS detection/prevention is on my longer-term TODO list.

gavinandresen committed:
Refactor, new CNode::FinalizeHeader method

I need this to write some unit tests for the
CNode::ReceiveMsgBytes() function.

gavinandresen committed:
Unit test for CNode::ReceiveMsgBytes

gavinandresen committed:
Allow per-message sanity checking when reading from wire

jgarzik commented on src/main.cpp:
```diff
@@ -3760,6 +3765,36 @@ std::string GetWarnings(const std::string& strFor)
+{
+    if (msg.hdr.nMessageSize > MAX_PROTOCOL_MESSAGE_LENGTH) {
+        LogPrint("net", "Oversized message from peer=%i, disconnecting", peer->GetId());
+        return false;
```
Harmonize the two LogPrint() calls.  No reason to omit strCommand diagnostic output on this first LogPrint() here.

jgarzik commented on src/net.cpp:
```diff
@@ -2017,6 +2015,22 @@ void CNode::AbortMessage() UNLOCK_FUNCTION(cs_vSend)
+    memcpy(&nChecksum, &hash, sizeof(nChecksum));
+    assert(s.size () >= CMessageHeader::CHECKSUM_OFFSET + sizeof(nChecksum));
+    memcpy((char*)&s[CMessageHeader::CHECKSUM_OFFSET], &nChecksum, sizeof(nChecksum));
+
```
Code is correct.  However, it seems odd to put this in CNode. </nit>

jgarzik commented:
ut ACK modulo comments

gavinandresen commented:
Addressed @jgarzik comments -- FinalizeHeader moved to CNetMessage (I considered moving it to CNetHeader, but didn't want to make protocol.h dependent on CDataStream).  And unified the log message (and remembered to SanitizeString(command) at the very last moment).

sipa commented:
Concept ACK, though I would just pass a pointer to CNetMessage rather than a wrapper-and-then-auto-unwrapped boost reference object.

gavinandresen commented:
@sipa: it'll be std::cref() when we move to C++11... and passing a reference is, in my opinion, cleaner than passing a pointer since the routine doesn't have to worry about checking for NULL.

Diapolo commented on src/main.cpp:
```diff
@@ -3765,6 +3770,32 @@ std::string GetWarnings(const std::string& strFor)
+    ("verack", 0)
+    ;
+
+bool static SanityCheckMessage(CNode* peer, const CNetMessage& msg)
```
Why did you choose to not directly pass the NodeId? It seems it's the only field used in here.

Diapolo commented on src/net.cpp:
```diff
@@ -564,6 +562,22 @@ bool CNode::ReceiveMsgBytes(const char *pch, unsigned int nBytes)
+unsigned int CNetMessage::FinalizeHeader(CDataStream& s)
+{
+    // Set the size
+    unsigned int nSize = s.size() - CMessageHeader::HEADER_SIZE;
```
``s.size()`` returns ``size_t``, so we have ``unsigned int = size_t - int``... is this correct?

gavinandresen commented:
@dgenr8 is seeing peer disconnections due to this code (it is part of the BIP101 pull request), especially with 'addr' messages. I need to investigate why.

And I like detecting misbehavior early, but this should work with the existing Misbehaving() logic. I think I see a path to get there:

1) Refactor mapNodeState so it is protected by it's own lock instead of cs_main. 
2) Rework this pull request to call Misbehaving() for oversize messages; never immediately disconnect.

sipa commented:
@gavinandresen 1000 addr messages is 3 + 1000*30 bytes, as the serialized vector length for sizes over 253 takes 3 bytes,

gavinandresen committed:
Refactor: protect mapNodeState with its own lock

Encapsulate mapNodeState in a smart-pointer class with its own lock.

Why? So Misbehaving() can be called from the sanity-check-a-message code without holding cs_main.

And to get better exception safety (the smart-pointer approach gives RAII semantics).

And because protecting fewer things with cs_main is a good idea.

Tested by compiling with -DDEBUG_LOCKORDER, running all of the qa/rpc-tests, and running a node on the main network overnight.

gavinandresen commented:
Reworked the mapNodeState encapsulation into a RAII NodeStatePtr smart-pointer class.

That allowed me to change the SanityCheckMessage() callback match the old behavior: if a peer send a message larger than 2MB, immediately drop their connection. If they send a message bigger than expected but less than 2MB, call Misbehaving(20) (which, by default, gets them dropped/banned if they send more than 5 oversize messages).

And I removed the addr/inv/getdata checks-- those messages already check for message length and call Misbehaving() if passed vectors that are too big.

mruddy commented:
I believe that the "version" message can be limited to 344 bytes. Right now I can send the full 2MB and it's accepted. The "filterclear" and "reject" messages need lower limits too as 2MB works for them with no penalty. Then that leaves the obvious final case in ProcessMessage that other fictitious/junk commands can be used with a 2MB payload to hog bandwidth.

I came looking for a ticket with a fix like this ticket implements because, I was playing around on my local node with a script that basically connected, sent a good "version" message (actually I used 2MB, but that is easy to fix with this ticket), then looped sending 2MB junk command messages. That might eventually timeout.If so, my script would have eventually had to maybe add in some pong and verack handling, but nothing difficult.

Maybe there's another ticket for doing something about the junk case that I didn't find yet too?

jgarzik commented:
concept ACK

MarcoFalke commented:
@gavinandresen Travis failed with `test_bitcoin: key.cpp:198: void ECC_Start(): Assertion `secp256k1_context == __null' failed.` This seems a temporary outage? Mind to run travis again?

laanwj commented:
Closing due to inactivity, feel free to reopen if you pick this up again.
